daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala

```diff
@@ -285,4 +285,32 @@ object Ref {
   type TypeSynName = Identifier
   val TypeSynName = Identifier
 
+  /*
+     max size when converted to string: 3068 ASCII chars
+   */
+  final case class QualifiedChoiceName(interfaceId: Option[Identifier], choiceName: ChoiceName) {
+    override def toString: String = interfaceId match {
+      case None => choiceName
+      case Some(ifaceId) => "#" + ifaceId.toString + "#" + choiceName
+    }
+  }
+
+  object QualifiedChoiceName {
+    def fromString(s: String): Either[String, QualifiedChoiceName] =
+      if (s.startsWith("#")) {
+        val i = s.indexOf('#', 1)
+        if (i < 0)
+          Left("Cannot parse qualified choice name")
+        else
+          for {
+            ifaceId <- Identifier.fromString(s.substring(1, i))
+            chName <- ChoiceName.fromString(s.substring(i + 1, s.length))
```

Comment:
Do we need to worry about the case where the choice name could be an empty string (e.g. `s == "#xxx#`)?

Reply:
Ah, this is not concern. the `s.substring(I + 1, s.length)` would fail in this case as we'd index past the string's length.

Reply:
In this case it does not fail, but return the empty string. 
I add a test case anyway, as this is an interesting corner cases.

Reply:
796f84eaf913c5cdff40cc1221445bbc2ae72fdb